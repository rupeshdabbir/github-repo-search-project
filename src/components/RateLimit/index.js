// @flow

import React from 'react';
import {
  StyleSheet,
  css,
} from 'aphrodite/no-important';

type Props = {
  label: string,
  reset: string,
  remaining: number,
  limit: number,
};

export function calculateRemainingColor(remaining: number, limit: number): string {
  if (remaining <= (limit / 4) * 1) {
    return 'RateLimit_low';
  }
  if (remaining <= (limit / 4) * 3) {
    return 'RateLimit_medium';
  }
  return 'RateLimit_high';
}

export default function RateLimit({label, remaining, limit, reset}: Props) {
  const styleKey = calculateRemainingColor(remaining, limit);
  return (
    <p
      className={`${css(styles.RateLimit)} u-flex`}
      title={`Limit resets at ${reset}`}
    >
      <span className={css(styles.RateLimit_label)}>{label}:</span>
      <span className={css(styles[styleKey])}>{remaining}</span>
    </p>
  );
}

RateLimit.defaultProps = {
  remaining: 0,
  limit: 0,
  reset: '',
};

const styles = StyleSheet.create({
  RateLimit: {
    fontSize: 14,
  },

  RateLimit_label: {
    marginRight: 3,
  },

  RateLimit_high: {
    color: '#10ce65',
  },

  RateLimit_medium: {
    color: '#f5f758',
  },

  RateLimit_low: {
    color: '#d6495d',
  },
});
