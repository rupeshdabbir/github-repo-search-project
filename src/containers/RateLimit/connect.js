// @flow

import { connect } from 'react-redux';
import get from 'lodash/fp/get';
import format from 'date-fns/format';

function formatDate(timestamp: number): string | void {
  if (!timestamp) {
    return undefined;
  }
  return format(new Date(timestamp * 1000), 'DD-MM-YYYY HH:mm');
}

export function mapStateToProps(state: Object): Object {
  return {
    coreLimit: get('rateLimit.core.limit', state),
    searchLimit: get('rateLimit.search.limit', state),
    searchRemaining: get('rateLimit.search.remaining', state),
    searchReset: formatDate(get('rateLimit.search.reset', state)),
  };
}

export default connect(mapStateToProps);
