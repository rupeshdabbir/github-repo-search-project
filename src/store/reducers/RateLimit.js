import assignAll from 'lodash/fp/assignAll';

const initialState = {
  core: {},
  search: {},
};

export default function rateLimitReducer(state: Object = initialState, action: Object) {
  switch (action.type) {

    case 'API_RATE_LIMIT_SUCCESS':
      return assignAll([
        state,
        {
          core: action.payload.core,
          search: action.payload.search,
        },
      ]);

    case 'API_RATE_LIMIT_FAILURE': {
      if (!action.error) {return state;}
      return assignAll([
        state,
        {
          payload: action.payload,
        },
      ]);
    }

    default:
      return state;

  }
}
