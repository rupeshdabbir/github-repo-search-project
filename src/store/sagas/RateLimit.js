import { put, take, takeLatest, select, call } from 'redux-saga/effects';
import keys from 'lodash/fp/keys';
import map from 'lodash/fp/map';
import get from 'lodash/fp/get';
import every from 'lodash/fp/every';
import flow from 'lodash/fp/flow';
import castArray from 'lodash/fp/castArray';
import api from 'store/api';

const isCachedAction = flow(castArray, every(get('meta.fromCache')));

export function* getRateLimit(actions) {
  if (isCachedAction(actions)) {
    const latestRateLimit = yield select(get('cache.rateLimit.latest'));
    yield put({
      payload: latestRateLimit,
      type: 'API_RATE_LIMIT_SUCCESS',
    });
    return;
  }

  try {
    const rateLimit = yield call(api.get, 'rate_limit');
    yield put({
      payload: get('data.resources', rateLimit),
      type: 'API_RATE_LIMIT_SUCCESS',
    });
  } catch (err) {
    yield put({
      error: true,
      payload: err,
      type: 'API_RATE_LIMIT_FAILURE',
    });
  }
}

export function* watchGetRateLimit() {
  const types = ['REPOS_SUCCESS', 'FOLLOWERS_SUCCESS', 'PROFILE_SUCCESS'];

  yield [combineLatest(types, getRateLimit), takeLatest('API_RATE_LIMIT_REQUEST', getRateLimit)];
}

/**
 * Tracks an array of action types and only calls the saga
 * when they all fire.
 * Resets once all actions fire
 *
 * Modified from https://github.com/jhewlett/redux-saga-combine-latest
 */
function* combineLatest(types, saga) {
  let actions = {};
  const getActions = map((type) => actions[type]);

  while (true) {
    const action = yield take(types);
    actions[action.type] = action;
    if (keys(actions).length === types.length) {
      yield saga(getActions(types));
      actions = {};
    }
  }
}
