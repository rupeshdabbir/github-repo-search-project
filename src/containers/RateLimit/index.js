import React, { Component } from 'react';
import { css, StyleSheet } from 'aphrodite/no-important';
import 'suitcss-utils-flex/lib/flex.css';
import RateLimit from 'components/RateLimit';
import isUndefined from 'lodash/fp/isUndefined';
import connect from './connect';

type Props = {
  coreLimit: ?number,
  dispatch: Function,
  searchLimit: ?number,
  searchRemaining: ?number,
  searchReset: ?string,
};

export class RateLimitContainer extends Component {
  constructor(props: Props) {
    super(props);
    const { coreLimit, searchLimit, dispatch } = props;

    if (isUndefined(coreLimit) || isUndefined(searchLimit)) {
      dispatch({ type: 'API_RATE_LIMIT_REQUEST' });
    }
  }

  render() {
    const { searchLimit, searchRemaining, searchReset } = this.props;

    return (
      <div className="u-flex u-flexAlignItemsCenter">
        <p className={css(styles.RateLimitContainer_label)}>API requests</p>
        <div className={`${css(styles.RateLimitContainer)} u-flex`}>
          <div className={css(styles.RateLimitContainer_item)}>
            <RateLimit label="Search" remaining={searchRemaining} limit={searchLimit} reset={searchReset} />
          </div>
        </div>
      </div>
    );
  }
}

const styles = StyleSheet.create({
  RateLimitContainer: {
    backgroundColor: '#333',
    color: '#eee',
    padding: '5px 8px',
    borderRadius: 5,
  },

  RateLimitContainer_label: {
    fontSize: 14,
    marginRight: 10,
  },

  RateLimitContainer_item: {
    ':first-child': {
      marginRight: 5,
    },

    ':last-child': {
      marginLeft: 5,
    },
  },
});

export default connect(RateLimitContainer);
