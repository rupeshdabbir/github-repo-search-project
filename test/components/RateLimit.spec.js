import React from 'react';
import renderer from 'react-test-renderer';
import {
  StyleSheetTestUtils,
} from 'aphrodite';
import RateLimit from 'components/RateLimit';

describe('Component: RateLimit', () => {

  beforeAll(() => {
    StyleSheetTestUtils.suppressStyleInjection();
  });

  afterAll(() => {
    StyleSheetTestUtils.clearBufferAndResumeStyleInjection();
  });

  it('should render `high` when above 75% of the API calls remaining', () => {
    const component = renderer.create(
      <RateLimit
        label="test"
        reset="some date"
        remaining={76}
        limit={100}
      />
    );
    expect(component).toMatchSnapshot();
  });

  it('should render `medium` when below 75% of the API calls remaining', () => {
    const component = renderer.create(
      <RateLimit
        label="test"
        reset="some date"
        remaining={70}
        limit={100}
      />
    );
    expect(component).toMatchSnapshot();
  });

  it('should render `low` when below 25% of the API calls remaining', () => {
    const component = renderer.create(
      <RateLimit
        label="test"
        reset="some date"
        remaining={24}
        limit={100}
      />
    );
    expect(component).toMatchSnapshot();
  });

});
