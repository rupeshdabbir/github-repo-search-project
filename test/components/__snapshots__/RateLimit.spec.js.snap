// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: RateLimit should render \`high\` when above 75% of the API calls remaining 1`] = `
<p
  className="RateLimit_f9e8qy u-flex"
  title="Limit resets at some date"
>
  <span
    className="RateLimit_label_1rcbo1g"
  >
    test
    :
  </span>
  <span
    className="RateLimit_high_9hrz9f"
  >
    76
  </span>
</p>
`;

exports[`Component: RateLimit should render \`low\` when below 25% of the API calls remaining 1`] = `
<p
  className="RateLimit_f9e8qy u-flex"
  title="Limit resets at some date"
>
  <span
    className="RateLimit_label_1rcbo1g"
  >
    test
    :
  </span>
  <span
    className="RateLimit_low_1aa01pl"
  >
    24
  </span>
</p>
`;

exports[`Component: RateLimit should render \`medium\` when below 75% of the API calls remaining 1`] = `
<p
  className="RateLimit_f9e8qy u-flex"
  title="Limit resets at some date"
>
  <span
    className="RateLimit_label_1rcbo1g"
  >
    test
    :
  </span>
  <span
    className="RateLimit_medium_zh4aso"
  >
    70
  </span>
</p>
`;
